dbms/src/Flash/Mpp/MPPTaskManager.cpp

```diff
@@ -27,13 +27,102 @@ namespace DB
+{
+    const auto & meta = request->sender_meta();
+    MPPTaskId id{meta.start_ts(), meta.task_id()};
+    Int64 sender_task_id = meta.task_id();
+    Int64 receiver_task_id = request->receiver_meta().task_id();
+
+    std::unique_lock lock(mu);
+    auto query_it = mpp_query_map.find(id.start_ts);
+    if (query_it == mpp_query_map.end())
+    {
+        if (!call_data->isWaitingTunnelState())
+        {
+            /// if call_data is in new_request state, then put it to waiting tunnel state
+            auto query_set = addMPPQueryTaskSet(id.start_ts);
+            auto & alarm = query_set->alarms[sender_task_id][receiver_task_id];
+            call_data->setToWaitingTunnelState();
+            alarm.Set(cq, Clock::now() + std::chrono::seconds(10), call_data);
+            return {nullptr, ""};
+        }
+        else
+        {
+            return {nullptr, fmt::format("Can't find task [{},{}] within 10 s.", id.start_ts, id.task_id)};
+        }
+    }
+    auto query_set = query_it->second;
+
+    if (!query_set->isInNormalState())
+    {
+        /// if the query is aborted, return true to stop waiting timeout.
+        LOG_WARNING(log, fmt::format("Query {} is aborted, all its tasks are invalid.", id.start_ts));
+        /// meet error
+        return {nullptr, query_it->second->error_message};
+    }
+    auto it = query_set->task_map.find(id);
+    if (it == query_set->task_map.end())
+    {
+        if (!call_data->isWaitingTunnelState())
+        {
+            /// task not found, use alarm
+            auto & alarm = query_set->alarms[sender_task_id][receiver_task_id];
+            call_data->setToWaitingTunnelState();
+            alarm.Set(cq, Clock::now() + std::chrono::seconds(10), call_data);
+            return {nullptr, ""};
+        }
+        else
+        {
+            if (query_it->second->task_map.empty())
+            {
+                /// if the query task set has no mpp task, it has to be removed if there is no alarms left,
+                /// otherwise the query task set itself may be left in MPPTaskManager forever
+                auto task_alarm_it = query_set->alarms.find(sender_task_id);
+                if (task_alarm_it != query_set->alarms.end())
+                {
+                    task_alarm_it->second.erase(receiver_task_id);
+                    if (task_alarm_it->second.empty())
+                    {
+                        query_set->alarms.erase(task_alarm_it);
+                        if (query_set->alarms.empty())
+                        {
+                            /// remove the mpp query task set if it is the last alarm
+                            removeMPPQueryTaskSet(id.start_ts, false);
+                            cv.notify_all();
+                        }
+                    }
+                }
+            }
+            return {nullptr, fmt::format("Can't find task [{},{}] within 10 s.", id.start_ts, id.task_id)};
```

Comment:
Could we merge this logic with the code between line 64-75? 
BTW, code between line 100 and 119 can be moved before `query_set->task_map.find(id)`.